### PR TITLE
add support for secret as an async function call to allow JWKS integration

### PR DESCRIPTION
**Pull request by dunnock**
Hi, auth0 is forcing users to switch to RSA algorithm with their new version, unfortunately they have integration only for Hapi and Express. I have implemented based on express implementation as per this example https://github.com/auth0/node-jwks-rsa/tree/master/examples/express-demo, tests and doc included. Can you please accept this PR?

Thanks, Maxim

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -385,7 +385,29 @@ describe('success tests', () => {
         .expect(validUserResponse)
         .end(done);
   });
-  
+
+  it('should work if secret is provided by secret callback', done => {
```
Comment (sdd) [inline test/test.js]: We also need tests for two more scenarios:

1) Secret callback returns an error
2) Secret callback returns a valid secret but authentication fails

**Comment by sdd** [comment]
I've made a start on a branch of `node-jwks-rsa` that includes an adapter for `koa-jwt` and an example - 

https://github.com/sdd/node-jwks-rsa/blob/koa-jwt-integration/src/integrations/koa.js
https://github.com/sdd/node-jwks-rsa/blob/koa-jwt-integration/examples/koa-demo/server.js

Nothing tested yet though. Will work on it later this evening. :+1:

**Comment by dunnock** [comment]
Thanks, will do

**Comment by dunnock** [comment]
weird, I did not change yarn.lock, any idea why travis might fail? `The command "eval yarn --frozen-lockfile " failed. Retrying, 2 of 3.`

File: lib/rsa.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const jwt = require('jsonwebtoken');
+
+module.exports = async (ctx, secretCallback, token) => {
+  const dtoken = jwt.decode(token, { complete: true });
```
Comment (dunnock) [inline lib/rsa.js]: fixed

File: lib/rsa.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+  return new Promise((resolve, reject) => {
+    secretCallback(ctx.req, dtoken.header, dtoken.payload, (err, secret) => {
```
Comment (dunnock) [inline lib/rsa.js]: fixed to the provided implementation

File: lib/index.js [hunk 4]
```diff
@@ -1,9 +1,14 @@
 'use strict';
 const unless = require('koa-unless');
 const verify = require('./verify');
+const getRSASecret = require('./rsa');
```
Comment (dunnock) [inline lib/index.js]: done

File: lib/index.js [hunk 5]
```diff
@@ -1,9 +1,14 @@
...
 const resolveAuthHeader = require('./resolvers/auth-header');
 const resolveCookies = require('./resolvers/cookie');
 
+function isFunction(object) {
```
Comment (dunnock) [inline lib/index.js]: fixed

File: test/test.js (hunk 1 above)
Comment (dunnock) [inline test/test.js]: done

**Comment by dunnock** [comment]
yarn fails since there are outdated deps:
```
-koa@2.0.1:
-  version "2.0.1"
-  resolved "https://registry.yarnpkg.com/koa/-/koa-2.0.1.tgz#c15a0da4d62554c058ae881e76992e3defe45bd8"
+koa@^2.0.1:
+  version "2.2.0"
+  resolved "https://registry.yarnpkg.com/koa/-/koa-2.2.0.tgz#b055933187849d540ad8b9f731baaa4be97c652d"
```

File: test/test.js [hunk 6]
```diff
@@ -386,16 +455,16 @@ describe('success tests', () => {
         .end(done);
   });
 
-  it('should work if secret is provided by secret callback', done => {
+  it('should work if secret is provided by deffered provider', done => {
     const validUserResponse = res => res.body.foo !== 'bar' && "Wrong user";
 
     const secret = 'shhhhhh';
-    const secretCb = (req, header, payload, callback) => callback(null, secret);
+    const provider = ({alg, kid}) => new Promise(resolve=>resolve(secret));
```
Comment (sdd) [inline test/test.js]: You can use `new Promise.resolve(secret)`here instead

File: test/test.js [hunk 7]
```diff
@@ -231,6 +231,75 @@ describe('failure tests', () => {
...
+      .expect('Invalid token - Not supported\n')
+      .end(done);
+  });
+
+
+  it('should throw if deffered provider reject', done => {
+    const validUserResponse = res => res.body.foo !== 'bar' && "Wrong user";
+
+    const secret = 'shhhhhh';
+    const provider = ({alg, kid}) => new Promise((_,reject)=>reject(new Error("Not supported")));
```
Comment (sdd) [inline test/test.js]: `new Promise.reject(new Error('Not supported'))` would be better here.

File: lib/jwks.js [hunk 8]
```diff
@@ -0,0 +1,7 @@
+const jwt = require('jsonwebtoken');
```
Comment (sdd) [inline lib/jwks.js]: `const { decode } = require('jsonwebtoken');` Would be a little nicer here - also the `console.log` needs removing :-)

File: lib/index.js [hunk 9]
```diff
@@ -32,9 +28,9 @@ module.exports = (opts = {}) => {
         }
 
         try {
-            if(isFunction(secret)) {
-              // secret is RSA callback (e.g. jwks-rsa.expressJwtSecret)
-              secret = await getRSASecret(ctx, secret, token);
+            if(typeof secret === 'function') {
+              // secret is RSA provider
```
Comment (sdd) [inline lib/index.js]: I don't think we even need a comment here. Could we name the function getSecret? This interface may be used contexts other than JWKS.

File: test/test.js [hunk 10]
```diff
@@ -231,6 +231,75 @@ describe('failure tests', () => {
       .end(done);
   });
 
+
+  it('should throw if deffered provider throws', done => {
```
Comment (sdd) [inline test/test.js]: I think when I suggested this test, I envisaged testing an error thrown directly, rather than wrapped in a promise. But Ive now remembered that since the location where this is called uses `await`, both a thrown error and a rejected promise are handled the same and converted into a thrown error. So, this first test (lines 235 to 255) can be removed as it is functionally equivalent to the one below it.

File: test/test.js (hunk 6 above)
Comment (dunnock) [inline test/test.js]: fixed

File: README.md [hunk 11]
```diff
@@ -206,12 +206,12 @@ var publicKey = fs.readFileSync('/path/to/public.pub');
 app.use(jwt({ secret: publicKey }));
 ```
 
-RSA algorithm is also supported by providing jwks-rsa.expressJwtSecret callback in the
+RSA algorithm is also supported by providing jwks promise provider in the
```
Comment (sdd) [inline README.md]: Could we rephrase this to:



If the `secret` option is a function, this function is called for each JWT received in order to determine which secret is used to verify the JWT.

The signature of this function should be ` ({ alg, kid }) => [Promise(secret)]`, where `alg` and `kid` are the algorithm and key id fields from the decoded JWT header.

This option can be used to support JWKS (JSON Web Key Set) providers by using [node-jwks-rsa](https://github.com/auth0/node-jwks-rsa). For example:


```javascript
const { koaJwtSecret } = require('jwks-rsa');

app.use(jwt({
    secret: koaJwtSecret({
        jwksUri: 'https://sandrino.auth0.com/.well-known/jwks.json',
        cache: true,
        cacheMaxEntries: 5,
        cacheMaxAge: ms('10h'),
    }),
    audience: 'http://myapi/protected',	
    issuer:   'http://issuer'
});
```

File: test/test.js (hunk 7 above)
Comment (dunnock) [inline test/test.js]: fixed

File: lib/jwks.js (hunk 8 above)
Comment (dunnock) [inline lib/jwks.js]: done

File: lib/index.js (hunk 9 above)
Comment (dunnock) [inline lib/index.js]: fixed, also renamed jwks.js to get-secret.js and pass whole header to the secret provider to make it more general, that will be destructured to {alg, kid} in jwks handler.

File: test/test.js (hunk 10 above)
Comment (dunnock) [inline test/test.js]: removed

**Comment by dunnock** [comment]
updated yarn.lock (just `yarn`)

File: README.md (hunk 11 above)
Comment (dunnock) [inline README.md]: done!

Comment (dunnock) [inline README.md]: I've changed middle paragraph a little bit to correspond to previous commit:

The signature of this function should be (header) => [Promise(secret)], where
header is token header. For instance to support JWKS token header should contain
alg and kid: algorithm and key id fields respectively.

File: README.md [hunk 12]
```diff
@@ -206,6 +206,29 @@ var publicKey = fs.readFileSync('/path/to/public.pub');
 app.use(jwt({ secret: publicKey }));
 ```
 
+If the secret option is a function, this function is called for each JWT received in
```
Comment (sdd) [inline README.md]: Sorry to be pedantic :-) This should be \`secret\` rather than secret. And `node-jwks-rsa` as a link, ie \[node-jwks-rsa\]\(https://github.com/auth0/node-jwks-rsa) so it looks like [node-jwks-rsa](https://github.com/auth0/node-jwks-rsa).

Once that is done, I think it is ready - thanks for your quick response this morning!

Comment (dunnock) [inline README.md]: Sorry, missed that, fixed couple more places

**Comment by sdd** [comment]
Looks good.

I'm nearly done on the PR for `node-jwks-rsa`, didn't quite have enough time on the train this morning to finish. Once [my tests](https://github.com/sdd/node-jwks-rsa/blob/koa-jwt-integration/tests/koa.tests.js) are complete, and the docs updated, I'll raise the PR with them.

I'll merge this now and publish to npm later today.

 thanks @dunnock!

**Comment by sdd** [comment]
This has now been published to npm as v3.1.0.

See (here)[https://github.com/koajs/jwt/releases/tag/v3.1.0] for release notes.
